### ISSUE TITLE
use of ::pe_version fails puppet run when strict_variables is enabled on foss puppet

### ISSUE THREAD

**Issue by smithtrevor** [acidprime/r10k#160]
The logic here uses the pe_version fact which doesn't exist on foss.  This throws 'Unknown variable' error.

https://github.com/acidprime/r10k/blob/master/manifests/install.pp#L67

**Comment by ghoneycutt** [comment]
@acidprime What do you think about pe_version in stdlib to prevent this?

**Comment by rnelson0** [comment]
@smithtrevor What version of Puppet are you using? With 3.7.5 FOSS this compiles properly.
```
[root@puppet ~]# puppet --version
3.7.5
[root@puppet ~]# puppet apply -e "unless versioncmp($::pe_version, '3.8.0') >= 0 { notify {'less than 3.8.0': } }"
Notice: Compiled catalog for puppet.nelson.va in environment production in 0.06 seconds
Notice: less than 3.8.0
Notice: /Stage[main]/Main/Notify[less than 3.8.0]/message: defined 'message' as 'less than 3.8.0'
Notice: Finished catalog run in 0.14 seconds
```

**Comment by smithtrevor** [comment]
@rnelson0 Originally it was on 3.7.5 and I get the same result on 3.7.3.  it's actually only an issue when strict_variables is set to true.  

```
[vagrant@cent7 ~]$ puppet --version
3.7.5
[vagrant@cent7 ~]$ puppet apply -e "unless versioncmp($::pe_version, '3.8.0') >= 0 { notify {'less than 3.8.0':} }"  --strict_variables true
Error: Undefined variable "::pe_version"; Undefined variable "pe_version" at line 1 on node cent7.local
Error: Undefined variable "::pe_version"; Undefined variable "pe_version" at line 1 on node cent7.local
```

A possible resolution to this would be to use getvar() in the conditional to test for existence of the fact prior to versioncmp()

```
[vagrant@cent7 ~]$ puppet apply -e "unless getvar('::pe_version') and versioncmp($::pe_version, '3.8.0') >= 0 { notify {'less than 3.8.0':} }"  --strict_variables true
Notice: Compiled catalog for cent7.local in environment production in 0.02 seconds
Notice: less than 3.8.0
Notice: /Stage[main]/Main/Notify[less than 3.8.0]/message: defined 'message' as 'less than 3.8.0'
Notice: Finished catalog run in 0.01 seconds
```

This results in the same behavior as the current implementation and makes more sense than my previous attempt with checking if $provider was set to 'pe_gem'.  If that approach makes sense I can update my PR.

pe_version is also used in webhook.pp and install/pe_gem.pp. in webhook.pp it's inside logic that checks if '$::is_pe == true' which is another fact I don't have, facter 2.4.3.  So it might make more sense to check for the fact existence and set a variable to the fact value or a sane default in params or something and use that variable throughout the module, rather than checking everywhere it's used.

**Comment by smithtrevor** [comment]
looks like using getvar() still fails under future parser with strict_variables

```
[vagrant@cent7 ~]$ puppet apply -e "unless getvar('::pe_version') and versioncmp($::pe_version, '3.8.0') >= 0 { notify {'less than 3.8.0':} }"  --strict_variables true --parser future
Error: Evaluation Error: Error while evaluating a Function Call, uncaught throw :undefined_variable at line 1:8 on node cent7.local
```

also fails under puppet 4 with strict_variables

```
[vagrant@cent7 ~]$ /opt/puppetlabs/bin/puppet --version
4.0.0
[vagrant@cent7 ~]$ /opt/puppetlabs/bin/puppet apply -e "unless getvar('::pe_version') and versioncmp($::pe_version, '3.8.0') >= 0 { notify {'less than 3.8.0':} }" --modulepath=/etc/puppet/environments/production/modules/ --strict_variables true
Error: Evaluation Error: Error while evaluating a Function Call, uncaught throw :undefined_variable at line 1:8 on node cent7.local
```

I guess the only viable option is to create the fact somewhere if you're using future parser or puppet 4.0.0. and using strict_variables

**Comment by smithtrevor** [comment]
pe_version is already provided by stdlib however it doesn't return anything on foss

```
Facter.add("pe_version") do
  setcode do
    pe_ver = Facter.value("puppetversion").match(/Puppet Enterprise (\d+\.\d+\.\d+)/)
    pe_ver[1] if pe_ver
  end
end
```
https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/lib/facter/pe_version.rb

is_pe is properly set to false on foss, I missed this in my previous comment as it is present I was just looking at a system that didn't have the plugin sync'd.  

The following is pretty much the same as what's in the webhooks.pp file.

```
[vagrant@cent7 ~]$ puppet --version
3.7.5
[vagrant@cent7 ~]$ puppet apply -e "unless ($::is_pe == 'true' or $::is_pe == true)  and versioncmp($::pe_version, '3.8.0') >= 0 { notify {'less than 3.8.0':} }"  --strict_variables true --parser future
Notice: Compiled catalog for cent7.local in environment production in 0.72 seconds
Notice: less than 3.8.0
Notice: /Stage[main]/Main/Notify[less than 3.8.0]/message: defined 'message' as 'less than 3.8.0'
Notice: Finished catalog run in 0.31 seconds
[vagrant@cent7 ~]$ puppet apply -e "unless ($::is_pe == 'true' or $::is_pe == true)  and versioncmp($::pe_version, '3.8.0') >= 0 { notify {'less than 3.8.0':} }"  --strict_variables true 
Notice: Compiled catalog for cent7.local in environment production in 0.05 seconds
Notice: less than 3.8.0
Notice: /Stage[main]/Main/Notify[less than 3.8.0]/message: defined 'message' as 'less than 3.8.0'
Notice: Finished catalog run in 0.33 seconds
```

and under puppet 4

```
[vagrant@cent7 ~]$ /opt/puppetlabs/bin/puppet --version
4.0.0
[vagrant@cent7 ~]$ /opt/puppetlabs/bin/puppet apply -e "unless ($::is_pe == 'true' or $::is_pe == true) and versioncmp($::pe_version, '3.8.0') >= 0 { notify {'less than 3.8.0':} }" --modulepath=/etc/puppet/environments/production/modules/ --strict_variables true
Notice: Compiled catalog for cent7.local in environment production in 0.34 seconds
Notice: less than 3.8.0
Notice: /Stage[main]/Main/Notify[less than 3.8.0]/message: defined 'message' as 'less than 3.8.0'
Notice: Applied catalog in 0.02 seconds
```

**Comment by acidprime** [comment]
Given the differences in PE I am not really sure I can do this reliably until I drop support for older versions of PE. PRs welcome.

**Comment by smithtrevor** [comment]
@acidprime the fix for this was in #161 which you merged.  I failed to come back and close the issue.  Sorry about that.  

There may be some other places where this occurs in the module but this fixed the one I ran into.
